src/ImageSharp/Memory/Allocators/Internals/Utilities.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Diagnostics;
+#if SUPPORTS_BITOPERATIONS
+using System.Numerics;
+#endif
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace SixLabors.ImageSharp.Memory.Allocators.Internals
+{
+    /// <summary>
+    /// Utility methods for array pooling.
+    /// </summary>
+    internal static class Utilities
+    {
+#if !SUPPORTS_BITOPERATIONS
+        private static ReadOnlySpan<byte> Log2DeBruijn => new byte[32]
+        {
+            00, 09, 01, 10, 13, 21, 02, 29,
+            11, 14, 16, 18, 22, 25, 03, 30,
+            08, 12, 20, 28, 15, 17, 24, 07,
+            19, 27, 23, 06, 26, 05, 04, 31
+        };
+#endif
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int SelectBucketIndex(int bufferSize)
+        {
+            Debug.Assert(bufferSize >= 0, "Should be greater than 0.");
+
+            // Buffers are bucketed so that a request between 2^(n-1) + 1 and 2^n is given a buffer of 2^n
+            // Bucket index is log2(bufferSize - 1) with the exception that buffers between 1 and 16 bytes
+            // are combined, and the index is slid down by 3 to compensate.
+            // Zero is a valid bufferSize, and it is assigned the highest bucket index so that zero-length
+            // buffers are not retained by the pool. The pool will return the Array.Empty singleton for these.
+#if SUPPORTS_BITOPERATIONS
+            return BitOperations.Log2(((uint)bufferSize - 1) | 15) - 3;
+#else
+            return Log2SoftwareFallback(((uint)bufferSize - 1) | 15) - 3;
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int GetMaxSizeForBucket(int binIndex)
+        {
+            int maxSize = 16 << binIndex;
+            Debug.Assert(maxSize >= 0, "Should be greater than 0.");
+            return maxSize;
+        }
+
+#if !SUPPORTS_BITOPERATIONS
+        /// <summary>
+        /// Returns the integer (floor) log of the specified value, base 2.
+        /// Note that by convention, input value 0 returns 0 since Log(0) is undefined.
+        /// Does not directly use any hardware intrinsics, nor does it incur branching.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        private static int Log2SoftwareFallback(uint value)
```

Comment:
Jpeg encoder uses similar logic for internal operations, this could potentially be placed somewhere in `Numerics.cs` class for reusage.

P.S.
Jpeg needs to calculate minimum bitsize of a given number which can be calculated via Log2DeBruijn logic. Not really a game changer but would be nice to have these under single class & single test fixture.

Reply:
That uses `BitOperations.LeadingZeroCount` does it not?

Reply:
If `Lzcnt` intrinsic is supported - yes. But for fallback code it now uses LuT of 255 values with possible branching if value exceeds 255 only once - thus 16 *bits* is the maximum value vitsize this method would safely calculate. DeBruijn sequence is should be a bit faster and would eliminate the 16 bit maximum value constraint. Code would be a bit different, that table is what can be shared between jpeg & pool code. I'll open a PR with naming fix & algorithms after finishing my current open pr for testing.

Reply:
Oh wait I'm dumb, `BitOperations.LeadingZeroCount` already contains fallback code with deBruijn fallback implementation, totally forgot that we ended up using that instead of raw lzcnt intrinsic. Afaik that table is internal so yeah, no point in actually putting this code to `Numerics.cs` as only pool would use it.

Reply:
Your maths is better than mine! 😀

Cool I’ll leave that to you then. Thanks!

Reply:
No problem! I'll try to do it today so you could use that in this pr before merging.

Reply:
Lovely, thanks!